### PR TITLE
Dropdown profile field devas 1544

### PR DESCRIPTION
**Pull request by vincentfretin**

### CODE REVIEW

File: assembl/static2/js/app/components/login/signupForm.jsx [hunk 1]
```diff
@@ -107,16 +114,44 @@ class SignupForm extends React.Component<void, SignupFormProps, SignupFormState>
...
+                        name={field.identifier === 'CUSTOM' ? field.id : field.identifier.toLowerCase()}
+                        placeholder={field.required ? `${field.title}*` : field.title}
+                        onChange={this.handleInput}
+                        required={field.required}
+                      />
+                    </FormGroup>
+                  );
+                }
+
+                if (field.__typename === 'SelectField' && field.options) {
```
Comment (ayyazdaniaryan) [inline assembl/static2/js/app/components/login/signupForm.jsx]: There is a code smell with this check. The `SelectField` model hasn't made an option mandatory. So, an empty select field was selected with no option wouldn't render. The impact is non-deterministic as the fields could be very important, but they aren't rendered for use, or non-important fields just won't render.

File: assembl/static2/js/app/graphql/fragments/SelectField.graphql [hunk 2]
```diff
@@ -0,0 +1,22 @@
+#import "./LangString.graphql"
+
+fragment selectField on SelectField {
+  # fieldType
```
Comment (ayyazdaniaryan) [inline assembl/static2/js/app/graphql/fragments/SelectField.graphql]: Perhaps can remove this comment.

File: assembl/static2/js/app/pages/discussionAdmin.jsx [hunk 3]
```diff
@@ -270,6 +271,24 @@ const mapStateToProps: MapStateToProps<ReduxState, *, *> = ({
...
   const { sectionsById, sectionsHaveChanged, sectionsInOrder } = sections;
   const { profileOptionsHasChanged, textFieldsById } = profileOptions;
+  const textFields = textFieldsById
+    .map(textField => textField)
+    .valueSeq()
+    .toJS();
+  textFields.forEach((field) => {
+    const newField = field;
+    if (field.options) {
+      // convert options to array, and remove _hasChanged on option that is set by moveItemUp/Down.
```
Comment (ayyazdaniaryan) [inline assembl/static2/js/app/pages/discussionAdmin.jsx]: Why? Apollo doesn't like a key that shouldn't be there?

File: assembl/static2/js/app/utils/translations.js [hunk 4]
```diff
@@ -647,7 +649,13 @@ const Translations = {
       },
       profileOptions: {
         addTextField: "Ajouter un champ",
+        addSelectField: "Add a dropdown field",
```
Comment (ayyazdaniaryan) [inline assembl/static2/js/app/utils/translations.js]: English in `fr` locale

File: assembl/models/configurable_fields.py [hunk 5]
```diff
@@ -131,6 +131,72 @@ class TextField(AbstractConfigurableField):
...
+        Integer, ForeignKey(
+            SelectField.id, ondelete='CASCADE', onupdate='CASCADE'),
+        nullable=False, index=True)
+
+    select_field = relationship(
+        SelectField, foreign_keys=(select_field_id,),
+        backref=backref("options", order_by="SelectFieldOption.order", cascade="all, delete-orphan"))
+    label = relationship(
+        LangString, foreign_keys=(label_id,),
+        backref=backref("label_of_option", lazy="dynamic"),
```
[the remark this answers is not in the document]
Comment (vincentfretin) [reply assembl/models/configurable_fields.py]: Yes, it's the backref for a langstring.

File: assembl/static2/js/app/components/administration/discussion/textField.jsx [hunk 6]
```diff
@@ -37,44 +39,48 @@ const TextField = ({
...
-          </Button>
+        <OverlayTrigger placement="top" overlay={required ? textFieldToggleOptionalTooltip : textFieldToggleRequiredTooltip}>
+          {/* overlaytrigger does not seem to work with SwitchButton so we add a span... */}
+          <span>
+            <SwitchButton name={`required-switch-${id}`} checked={required} onChange={toggleRequired} />
+          </span>
         </OverlayTrigger>
       )}
+      <div className="flex">
+        {!isLast ? (
```
[the remark this answers is not in the document]
Comment (vincentfretin) [reply assembl/static2/js/app/components/administration/discussion/textField.jsx]: We have the same stuff in other part of the administration yes.

File: assembl/static2/js/app/components/administration/discussion/selectFieldOptions.jsx [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+  moveOptionDown,
+  moveOptionUp,
+  editLocale,
+  fieldId,
+  options
+}: Props) => {
+  const confirmOptionDeletion = (id) => {
+    const modalTitle = <Translate value="administration.confirmSelectFieldOptionDeletionTitle" />;
+    const body = <Translate value="administration.confirmSelectFieldOptionDeletion" />;
+    const footer = [
```
[the remark this answers is not in the document]
Comment (vincentfretin) [reply assembl/static2/js/app/components/administration/discussion/selectFieldOptions.jsx]: As you see, yes, it's an array of one element. Our modal API take an array of buttons. I don't understand your question.

File: assembl/static2/js/app/components/login/signupForm.jsx [hunk 8]
```diff
@@ -80,6 +81,12 @@ class SignupForm extends React.Component<void, SignupFormProps, SignupFormState>
     inputHandler(this, e);
   }
 
+  handleSelectChange = (e) => {
+    const name = e.target.name;
+    const value = e.target.value ? [e.target.value] : null;
+    this.setState(() => ({ [name]: value }));
```
[the remark this answers is not in the document]
Comment (vincentfretin) [reply assembl/static2/js/app/components/login/signupForm.jsx]: Yes, name here is actually the id prop `id={field.id}`

File: assembl/static2/js/app/components/administration/discussion/selectFieldOptions.jsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (ayyazdaniaryan) [reply assembl/static2/js/app/components/administration/discussion/selectFieldOptions.jsx]: Good, it was just a question of whether this is an array of 1 item or more. Thanks!

File: assembl/static2/js/app/components/login/signupForm.jsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (ayyazdaniaryan) [reply assembl/static2/js/app/components/login/signupForm.jsx]: Yeah, but why is the `name` wrapped in an an array?

File: assembl/static2/js/app/pages/discussionAdmin.jsx (hunk 3 above)
Comment (vincentfretin) [reply assembl/static2/js/app/pages/discussionAdmin.jsx]: exactly, _hasChanged is not part of SelectFieldOptionInput

File: assembl/static2/js/app/utils/translations.js (hunk 4 above)
Comment (vincentfretin) [reply assembl/static2/js/app/utils/translations.js]: I forgot to remove this.

File: assembl/static2/js/app/graphql/fragments/SelectField.graphql (hunk 2 above)
Comment (vincentfretin) [reply assembl/static2/js/app/graphql/fragments/SelectField.graphql]: done

File: assembl/static2/js/app/components/login/signupForm.jsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (ayyazdaniaryan) [reply assembl/static2/js/app/components/login/signupForm.jsx]: Ahh okay! This is what happens when I don't write ES6 code for a while

File: assembl/static2/js/app/components/login/signupForm.jsx (hunk 1 above)
Comment (vincentfretin) [reply assembl/static2/js/app/components/login/signupForm.jsx]: It will render and empty select.

**Pull request merged**